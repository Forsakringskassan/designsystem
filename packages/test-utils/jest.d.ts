export * from "./dist/types/jest";
