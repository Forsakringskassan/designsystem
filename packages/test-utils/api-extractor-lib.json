{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
