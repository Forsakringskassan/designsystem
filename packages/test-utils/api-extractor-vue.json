{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
