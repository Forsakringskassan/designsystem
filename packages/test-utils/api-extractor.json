{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
