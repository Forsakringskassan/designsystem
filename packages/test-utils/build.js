const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
