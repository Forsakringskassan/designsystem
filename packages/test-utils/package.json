{
  "name": "@fkui/test-utils",
  "version": "6.6.0",
  "description": "FKUI test utils for Jest",
  "keywords": [
    "fkui",
    "designsystem",
    "test",
    "jest",
    "vue",
    "cypress"
  ],
  "homepage": "https://designsystem.forsakringskassan.se/",
  "bugs": "https://github.com/Forsakringskassan/designsystem/issues",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Forsakringskassan/designsystem.git",
    "directory": "packages/test-utils"
  },
  "license": "MIT",
  "author": "Försäkringskassan",
  "sideEffects": [
    "./dist/cjs/jest.js",
    "./dist/esm/jest.js",
    "./src/matchers/augmentations.ts",
    "./src/matchers/index.ts"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/lib.d.ts",
      "import": "./dist/esm/lib.js",
      "require": "./dist/cjs/lib.js"
    },
    "./jest": {
      "types": "./dist/types/jest.d.ts",
      "import": "./dist/esm/jest.js",
      "require": "./dist/cjs/jest.js"
    },
    "./vue": {
      "types": "./dist/types/vue.d.ts",
      "import": "./dist/esm/vue.js",
      "require": "./dist/cjs/vue.js"
    }
  },
  "main": "dist/cjs/lib.js",
  "types": "dist/types/lib.d.ts",
  "files": [
    "dist",
    "jest.d.ts",
    "vue.d.ts"
  ],
  "scripts": {
    "prebuild": "rimraf dist temp && tsc",
    "build": "node build.js",
    "postbuild": "fk-api-extractor --patch-augmentations api-extractor.json api-extractor-*.json",
    "clean": "rimraf .jest-cache dist temp test-results",
    "prepack": "release-prepack --bundle --retain-scripts",
    "postpack": "release-postpack",
    "prepublishOnly": "release-prepublish --bundle --retain-scripts",
    "test": "jest"
  },
  "jest": {
    "preset": "@forsakringskassan/jest-config-vue"
  },
  "devDependencies": {
    "@vue/server-renderer": "3.5.13",
    "@vue/test-utils": "2.4.6",
    "esbuild": "0.25.3",
    "jest-environment-jsdom": "29.7.0",
    "vue": "3.5.13"
  },
  "peerDependencies": {
    "jest": "^26.0.0 || ^27.0.0 || ^28.0.0 || ^29.0.0",
    "vue": "^3.0.0"
  },
  "peerDependenciesMeta": {
    "vue": {
      "optional": true
    }
  },
  "engines": {
    "node": ">= 20",
    "npm": ">= 7"
  }
}
