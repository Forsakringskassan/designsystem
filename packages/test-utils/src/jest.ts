/* imported for side-effects only */
import "./matchers";

export * from "./lib";
