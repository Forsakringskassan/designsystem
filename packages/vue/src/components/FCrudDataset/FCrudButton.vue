<template>
    <button type="button" class="button button--small button--tertiary" @click="executeAction">
        <f-icon v-if="icon" class="button__icon" :name="iconName"></f-icon>
        <span v-if="!label" class="sr-only">
            <slot> {{ buttonText }} </slot>
        </span>
        <slot v-if="label"> {{ buttonText }} </slot>
    </button>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { FIcon } from "../FIcon";
import { TranslationMixin } from "../../plugins";
import { FCrudDatasetInjected, FCrudDatasetInterface } from "./FCrudDatasetInterface";

export default defineComponent({
    name: "FCrudButton",
    components: { FIcon },
    mixins: [TranslationMixin],
    props: {
        action: {
            type: String,
            required: true,
            validator(value: string) {
                return ["delete", "modify"].includes(value);
            },
        },
        icon: {
            type: Boolean,
            default: false,
        },
        item: {
            type: Object,
            required: true,
        },
        label: {
            type: Boolean,
            default: false,
        },
    },
    setup(): FCrudDatasetInterface {
        return FCrudDatasetInjected();
    },
    computed: {
        iconName(): string {
            if (this.action === "delete") {
                return "trashcan";
            } else {
                return "pen";
            }
        },
        buttonText(): string {
            if (this.action === "delete") {
                return this.$t("fkui.crud-button.delete", "Ta bort");
            } else {
                return this.$t("fkui.crud-button.modify", "Ã„ndra");
            }
        },
    },
    methods: {
        executeAction(): void {
            if (this.action === "delete") {
                this.delete(this.item);
            } else {
                this.modify(this.item);
            }
        },
    },
});
</script>
