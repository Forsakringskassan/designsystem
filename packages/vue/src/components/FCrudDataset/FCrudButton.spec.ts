import "html-validate/jest";
import { VueWrapper, mount } from "@vue/test-utils";
import FCrudButton from "./FCrudButton.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FCrudButton, {
        attrs: { ...attrs },
        props: {
            action: "modify",
            item: { id: 1, name: "test" },
            ...props,
        },
        slots: { ...slots },
        global: {
            provide: {
                delete() {
                    /* do nothing */
                },
                modify() {
                    /* do nothing */
                },
            },
            stubs: ["FIcon"],
        },
    });
}

describe("snapshot", () => {
    it("should match snapshot when action = modify", () => {
        const wrapper = createWrapper({
            props: { action: "modify" },
        });
        expect(wrapper).toMatchSnapshot();
    });

    it("should match snapshot when action = delete", () => {
        const wrapper = createWrapper({
            props: { action: "delete" },
        });
        expect(wrapper).toMatchSnapshot();
    });

    it("should match snapshot when only icon is used", () => {
        const wrapper = createWrapper({
            props: { action: "modify", icon: true },
        });
        expect(wrapper).toMatchSnapshot();
    });

    it("should match snapshot when only label is used", () => {
        const wrapper = createWrapper({
            props: { action: "modify", label: true },
        });
        expect(wrapper).toMatchSnapshot();
    });

    it("should match snapshot when both label and icon is used", () => {
        const wrapper = createWrapper({
            props: { action: "modify", label: true, icon: true },
        });
        expect(wrapper).toMatchSnapshot();
    });
});

describe("html-validate", () => {
    it("should require action attribute", () => {
        const markup = /* HTML */ `
            <f-crud-dataset>
                <template #default>
                    <f-crud-button item="test"></f-crud-button>
                </template>
            </f-crud-dataset>
        `;
        expect(markup).toMatchInlineCodeframe(`
            "error: <f-crud-button> is missing required "action" attribute (element-required-attributes) at inline:4:22:
              2 |             <f-crud-dataset>
              3 |                 <template #default>
            > 4 |                     <f-crud-button item="test"></f-crud-button>
                |                      ^^^^^^^^^^^^^
              5 |                 </template>
              6 |             </f-crud-dataset>
              7 |
            Selector: f-crud-dataset > template > f-crud-button"
        `);
    });

    it("should require item attribute", () => {
        const markup = /* HTML */ `
            <f-crud-dataset>
                <template #default>
                    <f-crud-button action="modify"></f-crud-button>
                </template>
            </f-crud-dataset>
        `;
        expect(markup).toMatchInlineCodeframe(`
            "error: <f-crud-button> is missing required "item" attribute (element-required-attributes) at inline:4:22:
              2 |             <f-crud-dataset>
              3 |                 <template #default>
            > 4 |                     <f-crud-button action="modify"></f-crud-button>
                |                      ^^^^^^^^^^^^^
              5 |                 </template>
              6 |             </f-crud-dataset>
              7 |
            Selector: f-crud-dataset > template > f-crud-button"
        `);
    });

    it("should only allow actions modify and delete", () => {
        const markup = (action: string): string => /* HTML */ `
            <f-crud-dataset>
                <template #default>
                    <f-crud-button
                        action="${action}"
                        item="test"
                    ></f-crud-button>
                </template>
            </f-crud-dataset>
        `;

        expect(markup("modify")).toMatchInlineCodeframe(`""`);
        expect(markup("delete")).toMatchInlineCodeframe(`""`);
        expect(markup("foobar")).toMatchInlineCodeframe(`
            "error: Attribute "action" has invalid value "foobar" (attribute-allowed-values) at inline:5:33:
              3 |                 <template #default>
              4 |                     <f-crud-button
            > 5 |                         action="foobar"
                |                                 ^^^^^^
              6 |                         item="test"
              7 |                     ></f-crud-button>
              8 |                 </template>
            Selector: f-crud-dataset > template > f-crud-button"
        `);
    });

    it("should require f-crud-dataset as ancestor", () => {
        const markup = /* HTML */ `
            <f-crud-button action="modify" item="test"></f-crud-button>
        `;
        expect(markup).toMatchInlineCodeframe(`
            "error: <f-crud-button> element requires a <f-crud-dataset> ancestor (element-required-ancestor) at inline:2:14:
              1 |
            > 2 |             <f-crud-button action="modify" item="test"></f-crud-button>
                |              ^^^^^^^^^^^^^
              3 |
            Selector: f-crud-button"
        `);
    });
});
