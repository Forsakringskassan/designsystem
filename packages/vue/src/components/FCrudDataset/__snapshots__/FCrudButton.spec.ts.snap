// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should match snapshot when action = delete 1`] = `
<button type="button" class="button button--small button--tertiary">
  <!--v-if--><span class="sr-only">Ta bort</span>
  <!--v-if-->
</button>
`;

exports[`snapshot should match snapshot when action = modify 1`] = `
<button type="button" class="button button--small button--tertiary">
  <!--v-if--><span class="sr-only">Ändra</span>
  <!--v-if-->
</button>
`;

exports[`snapshot should match snapshot when both label and icon is used 1`] = `
<button type="button" class="button button--small button--tertiary">
  <f-icon-stub name="pen" library="f" class="button__icon"></f-icon-stub>
  <!--v-if-->Ändra
</button>
`;

exports[`snapshot should match snapshot when only icon is used 1`] = `
<button type="button" class="button button--small button--tertiary">
  <f-icon-stub name="pen" library="f" class="button__icon"></f-icon-stub><span class="sr-only">Ändra</span>
  <!--v-if-->
</button>
`;

exports[`snapshot should match snapshot when only label is used 1`] = `
<button type="button" class="button button--small button--tertiary">
  <!--v-if-->
  <!--v-if-->Ändra
</button>
`;
